### PR TITLE
Tuned various battle mechanics

### PR DESCRIPTION
**Pull request by Domimmo314**
### BROAD
 1. Implemented the walk delay: on the frame entities resume walking status, they don't move, only visually transition to walking animation
 2. Split entity update loop to separate movement from attack: dispatching an entity in one go makes that entity unable to respond (= initiate an attack) to changes in positions of subsequent moved entities. One of the most absurd result of this is an entity with a smaller range can have a frame advantage on a bigger ranged one
 3. Internally ordered entities {[cats], [enemies]}. Because of drawing levels, layers order was being used for general updating too. This was biased. The question is whether in BC for the update loop cats and enemies are mixed together in a single list. If this is were to be case then the finer order of all the entities would make a difference and the game would be much more sensible to chances, if instead factions are not mixed up the inside-faction order doesn't matter because entities react solely to opposite-side entities. One telling experience for this is that moving cats always move in the frame they initiate an attack, while enemy can initiate an attack without moving. But there are also other effects of this clean cut, like the following.
 4. Made possible for cats freeze to block targeted enemies from progressing their attack in the frame they're being frozen: imperfect perfect freeze is now possible
 5. Changed the target for single-attackers from closest to their position to furthest behind into their hitbox
 6. Changed wave stoppers from blocking only the wave that would hit them to block also the one behind
 7. Adjusted waves and surges timings
 8. Adjusted limit position formula for cats / enemies / boss
 9. Adjusted frame-0 and frame-1 spawn enemies and respawn timings

### SPECIFIC
10. Tightened cannons, all of them, ranges, timings, lingering
11. Fixed zombie cannon freeze duration receiving zombie fruit boost
12. Fixed metal cannon damage rounding
13. Adjusted zombie burrowing/reviving transitions timings/interactions with cats
14. Adjusted self-destruct units post attack linger
15. Added summoned spirits enemy base limit
16. Fixed a problem with speed up combos yielding unit steps finer than .25 distance quantization

### VISUAL / NON-MECHANICS
17. Put spirits on same layer (0) so that same frame deaths share same height spirits
18. Cleaned up entity search (1-2-3 letters input get an exact search, >3 letters fuzzy search) 
19. Fixed dojo timer display and capped it to 100 min
20. Fixed app crashing when ordering by min position

### CODE REVIEW

File: battle/StageBasis.java [hunk 1]
```diff
@@ -143,12 +143,7 @@ public StageBasis(BattleField bf, EStage stage, BasisLU bas, int[] ints, long se
 		canon = new Cannon(this, nyc[0], nyc[1], nyc[2]);
 		conf = ints;
 
-		if(st.minSpawn <= 0 || st.maxSpawn <= 0)
-			respawnTime = 1;
-		else if(st.minSpawn == st.maxSpawn)
-			respawnTime = st.minSpawn;
-		else
-			respawnTime = st.minSpawn + (int) ((st.maxSpawn - st.minSpawn) * r.nextDouble());
+		respawnTime = 0;
```
Comment (iREDMe) [inline battle/StageBasis.java:146]: initial spawn time is based on min/max respawn time set for that stage (e.g. 1~60f for earthshaker) so i'd like that segment to still be there

File: battle/entity/Cannon.java [hunk 2]
```diff
@@ -181,75 +181,86 @@ public void update() {
...
+            int spe = 150;
             float p = b.ubase.pos - wid / 2f + spe;
             AttackCanon eatk = new AttackCanon(this, 0, traits, 0, proc, 0, 0, 1);
-            new ContExtend(eatk, p, wid, spe, 1, 31, 0, 9);
+            new ContExtend(eatk, p, wid, spe, 1, 32, 0, 9);
         } else if (id == 3) {
             // freeze canon
             traits.add(UserProfile.getBCData().traits.get(TRAIT_TOT));
-            duration = 1;
+            duration = 1 + 10;
```
Comment (iREDMe) [inline battle/entity/Cannon.java:218]: 11

File: battle/entity/Cannon.java [hunk 3]
```diff
@@ -181,75 +181,86 @@ public void update() {
...
             proc.STOP.time = (int) (b.b.t().getCannonMagnification(id, Data.BASE_TIME) * (100 + b.b.getInc(C_STOP)) / 100.0);
             int atk = (int) (b.b.t().getCanonAtk() * b.b.t().getCannonMagnification(id, Data.BASE_ATK_MAGNIFICATION) / 100.0);
-            int rad = NYRAN[3] / 2;
+            float rad = NYRAN[3] / 2;
             b.getAttack(new AttackCanon(this, atk, traits, 0, proc, pos - rad, pos + rad, duration));
         } else if (id == 4) {
             // water canon
             traits.add(UserProfile.getBCData().traits.get(TRAIT_TOT));
-            duration = 1;
+            duration = 1 + 10;
```
Comment (iREDMe) [inline battle/entity/Cannon.java:226]: 11

File: battle/entity/Cannon.java [hunk 4]
```diff
@@ -181,75 +181,86 @@ public void update() {
...
+            float p = (float) (b.ubase.pos - 332.5 + wid / 2);
             AttackCanon eatk = new AttackCanon(this, 0, traits, AB_ONLY | AB_ZKILL | AB_CKILL, proc, 0, 0, 1);
             new ContWaveCanon(new AttackWave(eatk.attacker, eatk, p, wid, WT_CANN | WT_WAVE), p, 5);
         } else if (id == 6) {
-            // barrier canon
-            // guessed hit box time
+            // blast canon
             traits.addAll(UserProfile.getAll(Identifier.DEF, Trait.class));
-            duration = 11;
+            duration = 1 + 10;
```
Comment (iREDMe) [inline battle/entity/Cannon.java:244]: 11

File: battle/entity/Entity.java [hunk 5]
```diff
@@ -2383,13 +2406,13 @@ else if (criti != 0)
 			if (crit > 0)
 				ans = (int) (ans * 0.01 * crit);
 			else if (crit < 0)
-				ans = (int) Math.ceil(health * crit / -100.0);
+				ans = (int) Math.max(1, health * crit / -100);
```
Comment (iREDMe) [inline battle/entity/Entity.java:2409]: Keep Math.ceil, the int cast only floors

File: battle/entity/Entity.java [hunk 6]
```diff
@@ -2383,13 +2406,13 @@ else if (criti != 0)
...
 			else if (crit < 0)
-				ans = (int) Math.ceil(health * crit / -100.0);
+				ans = (int) Math.max(1, health * crit / -100);
 			else
 				ans = ans > 0 ? 1 : 0;
 		else if (crit > 0)
 			ans = (int) (ans * 0.01 * crit);
 		else if (crit < 0)
-			ans = (int) Math.ceil(health * 0.001);
+			ans = (int) Math.max(1, health / 1000);
```
Comment (iREDMe) [inline battle/entity/Entity.java:2415]: Keep Math.ceil, the int cast only floors

**Review by iREDMe — CHANGES_REQUESTED** [review]

**Comment by Skwowwq** [comment]
Djdijwjw

**Comment by Skwowwq** [comment]
Auaiajxnsia

**Comment by Skwowwq** [comment]
hi

**Pull request closed without merging**